cpp/pagerank_module/algorithm/pagerank.cpp

```diff
@@ -0,0 +1,144 @@
+#include <algorithm>
+#include <numeric>
+#include <queue>
+
+#include "pagerank.hpp"
+
+namespace pagerank_alg {
+
+PageRankGraph::PageRankGraph(const std::uint64_t number_of_nodes, const std::uint64_t number_of_edges,
+                             const std::vector<std::pair<std::uint64_t, std::uint64_t>> &edges)
+    : node_count_(number_of_nodes),
+      edge_count_(number_of_edges),
+      out_degree_(std::vector<std::uint64_t>(number_of_nodes)) {
+  pagerank_util::AdjacencyList<std::uint64_t> in_neighbours;
+  in_neighbours.Init(number_of_nodes);
+  for (const auto &[from, to] : edges) {
+    // Because PageRank needs a set of nodes that point to a given node.
+    in_neighbours.AddAdjacentPair(from, to);
+    out_degree_[from] += 1;
+  }
+  for (std::size_t node_id = 0; node_id < number_of_nodes; node_id++) {
+    const auto &adjacent_nodes = in_neighbours.GetAdjacentNodes(node_id);
+    for (const auto &adjacent_node : adjacent_nodes) {
+      ordered_edges_.emplace_back(adjacent_node, node_id);
+    }
+  }
+}
+
+std::uint64_t PageRankGraph::GetNodeCount() const { return node_count_; }
+
+std::uint64_t PageRankGraph::GetEdgeCount() const { return edge_count_; }
+
+const std::vector<std::pair<std::uint64_t, std::uint64_t>> &PageRankGraph::GetOrderedEdges() const {
+  return ordered_edges_;
+}
+
+std::uint64_t PageRankGraph::GetOutDegree(const std::uint64_t node_id) const { return out_degree_[node_id]; }
+
+void CalculateOptimalBorders(const PageRankGraph &graph, const int number_of_threads,
+                             std::vector<std::uint64_t> *borders) {
+  for (int i = 0; i <= number_of_threads; i++) {
+    borders->push_back(static_cast<uint64_t>(i) * graph.GetEdgeCount() / number_of_threads);
+  }
```

Comment:
```suggestion
void CalculateOptimalBorders(const PageRankGraph &graph, const std::uint64_t number_of_threads,
                             std::vector<std::uint64_t> *borders) {
  for (std::uint64_t i = 0; i <= number_of_threads; i++) {
    borders->push_back(i * graph.GetEdgeCount() / number_of_threads);
  }
```
Since you are actually sending uint64_t in  lines 98-101:
```
  const std::uint64_t number_of_threads = std::thread::hardware_concurrency();

  std::vector<std::uint64_t> borders;
  CalculateOptimalBorders(graph, number_of_threads, &borders);
```